File: python-version/TicTacToe.py
import random
import copy

class TicTacToePlayer:
    # object representation for an AI Tic Tac Toe player.
    board = [[' ' for _ in range(3)] for _ in range(3)]
    pieces = ['X', 'O']

    def __init__(self):
        # Initializes an AI Tic Tac Toe player object
        self.my_piece = random.choice(self.pieces)
        self.opp = self.pieces[0] if self.my_piece == self.pieces[1] else self.pieces[1]

    # select a (row,col) space for the next move
    def make_move(self, state):
        best_value = float("-inf")
        best_state = None

        # get the best state of the board
        for successor in self.succ(state, self.my_piece):
            current_value = self.min_value(float("-inf"), float("inf"), successor, 1)
            if current_value > best_value:
                best_value = current_value
                best_state = successor

        move = []

        # Guard against None (can happen if board is already full)
        if best_state is None:
            return move

        # get the best move
        for i in range(3):
            for j in range(3):
                # if the space is empty and does not already exist in the board state
                if state[i][j] == ' ' and state[i][j] != best_state[i][j]:
                    # add it to the best move
                    (row, col) = (i, j)
                    move.insert(0, (row, col))
        return move

    def succ(self, state, piece):
        # Returns a list of all possible successor states from board state
        successors = []

        for row in range(3):
            for col in range(3):
                state_copy = copy.deepcopy(state)  # copy of the board state so that it does not change the original during searching
                if state_copy[row][col] == ' ':     # if the space is empty
                    state_copy[row][col] = piece    # add it to the board
                    successors.append(state_copy)   # add copied board state as a successor

        return successors

# minimax algorithm
    def max_value(self, a, b, state, depth):
        # return the maximum value of the state (+1)
        # check whether the state is a terminal state
        if self.game_value(state) != 0:
            return self.game_value(state)

        successors = self.succ(state, self.my_piece)
        if not successors:  # board is full, it's a draw
            return 0

        for successor in successors:
            a = max(a, self.min_value(a, b, successor, depth + 1))  # get the maximum alpha
            # alpha pruning
            if a >= b:
                return b

        return a

    def min_value(self, a, b, state, depth):
        # return the minimum value of the state (-1)
        # check whether the state is a terminal state
        if self.game_value(state) != 0:
            return self.game_value(state)

        successors = self.succ(state, self.opp)
        if not successors:  # board is full, it's a draw
            return 0

        for successor in successors:
            b = min(b, self.max_value(a, b, successor, depth + 1))  # get the minimum beta
            # beta pruning
            if a >= b:
                return a

        return b

# Game-related methods
    def opponent_move(self, move):
        if self.board[move[0][0]][move[0][1]] != ' ':  # validate the player's move to make sure the space is empty and ready to be filled
            raise Exception("Illegal move detected")
        self.place_piece(move, self.opp)

    def place_piece(self, move, piece):  # place the piece at the location on the board
        self.board[move[0][0]][move[0][1]] = piece

    def print_board(self):
        for row in range(len(self.board)):
            line = str(row + 1) + " "  # row
            for cell in self.board[row]:
                line += cell + " "     # spaces to be filled
            print(line)
        print("  1 2 3 \n")            # column

    # check the current state of the board whether there is a winning condition
    def game_value(self, state):
        # 1 if AI wins
        # -1 if player wins

        # Check rows '-'
        for row in range(3):
            if state[row][0] != ' ' and state[row][0] == state[row][1] == state[row][2]:
                return 1 if state[row][0] == self.my_piece else -1

        # Check columns '|'
        for col in range(3):
            if state[0][col] != ' ' and state[0][col] == state[1][col] == state[2][col]:
                return 1 if state[0][col] == self.my_piece else -1

        # Check diagonal '\'
        if state[0][0] != ' ' and state[0][0] == state[1][1] == state[2][2]:
            return 1 if state[0][0] == self.my_piece else -1

        # Check diagonal '/'
        if state[0][2] != ' ' and state[0][2] == state[1][1] == state[2][0]:
            return 1 if state[0][2] == self.my_piece else -1

        return 0  # no winner

# Sample gameplay loop for testing purposes
def main():
    print("Welcome to Tic Tac Toe Game!")
    ai = TicTacToePlayer()
    turn = 0
    total_pieces = 0  # keep track of the number of pieces on the board

    # while there is no winning condition and the board is not full
    while ai.game_value(ai.board) == 0 and total_pieces < 9:
        # if it is the ai's turn
        if ai.my_piece == ai.pieces[turn]:
            ai.print_board()                        # display the current state of the board
            move = ai.make_move(ai.board)           # call the make_move method to evaluate the best move
            ai.place_piece(move, ai.my_piece)       # call the place piece method to make the move
            print(f"{ai.my_piece} moved at {move[0]}")  # display the move taken
            total_pieces += 1                       # increment the total number of pieces
        # if it is the player's turn
        else:
            move_made = False                       # if the player hasn't made their move
            ai.print_board()                        # display the current state of the board
            print(f"{ai.opp}'s turn")
            while not move_made:
                player_move = input("Move (e.g. row,column): ")  # ask player for their move
                try:
                    row, col = map(int, player_move.split(','))
                except ValueError:
                    print("Invalid format. Use row,column (e.g. 2,3)")
                    continue
                if str(row) not in "123" or str(col) not in "123":  # validate input
                    print("Row and column must be between 1 and 3.")
                    continue
                try:
                    ai.opponent_move([(row - 1, col - 1)])  # make the move
                    move_made = True                         # set move_made to true to stop looping
                except Exception as e:
                    print(e)
            total_pieces += 1  # increment the total number of pieces
        turn += 1
        turn %= 2

    ai.print_board()
    if ai.game_value(ai.board) == 1:
        print("AI wins! Game over.")
    elif ai.game_value(ai.board) == -1:
        print("You win! Game over.")
    else:
        print("It's a draw! Game over.")

if __name__ == "__main__":
    main()

File: java-version/TicTacToe.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.Scanner;

public class TicTacToe {
    // ─── TicTacToePlayer (inner class) ───────────────────────────────────────
    static class TicTacToePlayer {

        private static final char[] PIECES = {'X', 'O'};

        char[][] board;
        char myPiece;
        char opp;

        TicTacToePlayer() {
            board = new char[3][3];
            for (int i = 0; i < 3; i++)
                for (int j = 0; j < 3; j++)
                    board[i][j] = ' ';

            myPiece = PIECES[new Random().nextInt(2)];
            opp = (myPiece == PIECES[0]) ? PIECES[1] : PIECES[0];
        }

        // ── make_move ─────────────────────────────────────────────────────────
        // Returns a list containing one (row, col) pair – the best move.
        List<int[]> makeMove(char[][] state) {
            double bestValue = Double.NEGATIVE_INFINITY;
            char[][] bestState = null;

            for (char[][] successor : succ(state, myPiece)) {
                double currentValue = minValue(Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY,
                                               successor, 1);
                if (currentValue > bestValue) {
                    bestValue = currentValue;
                    bestState = successor;
                }
            }

            List<int[]> move = new ArrayList<>();
            for (int i = 0; i < 3; i++) {
                for (int j = 0; j < 3; j++) {
                    if (state[i][j] == ' ' && bestState != null && state[i][j] != bestState[i][j]) {
                        move.add(0, new int[]{i, j});
                    }
                }
            }
            return move;
        }

        List<char[][]> succ(char[][] state, char piece) {
            List<char[][]> successors = new ArrayList<>();
            for (int row = 0; row < 3; row++) {
                for (int col = 0; col < 3; col++) {
                    if (state[row][col] == ' ') {
                        char[][] copy = deepCopy(state);
                        copy[row][col] = piece;
                        successors.add(copy);
                    }
                }
            }
            return successors;
        }

        // ── Minimax ───────────────────────────────────────────────────────────
        double maxValue(double a, double b, char[][] state, int depth) {
            if (gameValue(state) != 0) return gameValue(state);

            for (char[][] successor : succ(state, myPiece)) {
                a = Math.max(a, minValue(a, b, successor, depth + 1));
                if (a >= b) return b;  // alpha pruning
            }
            return a;
        }

        double minValue(double a, double b, char[][] state, int depth) {
            if (gameValue(state) != 0) return gameValue(state);

            for (char[][] successor : succ(state, opp)) {
                b = Math.min(b, maxValue(a, b, successor, depth + 1));
                if (a >= b) return a;  // beta pruning
            }
            return b;
        }

        void opponentMove(List<int[]> move) throws Exception {
            int row = move.get(0)[0];
            int col = move.get(0)[1];
            if (board[row][col] != ' ')
                throw new Exception("Illegal move detected");
            placePiece(move, opp);
        }

        void placePiece(List<int[]> move, char piece) {
            board[move.get(0)[0]][move.get(0)[1]] = piece;
        }

        void printBoard() {
            for (int row = 0; row < 3; row++) {
                StringBuilder line = new StringBuilder((row + 1) + " ");
                for (char cell : board[row])
                    line.append(cell).append(" ");
                System.out.println(line);
            }
            System.out.println("  1 2 3\n");
        }

        // Returns  1 if AI wins, -1 if player wins, 0 otherwise.
        int gameValue(char[][] state) {
            // Check rows
            for (int row = 0; row < 3; row++) {
                if (state[row][0] != ' ' &&
                    state[row][0] == state[row][1] && state[row][1] == state[row][2])
                    return (state[row][0] == myPiece) ? 1 : -1;
            }
            // Check columns
            for (int col = 0; col < 3; col++) {
                if (state[0][col] != ' ' &&
                    state[0][col] == state[1][col] && state[1][col] == state[2][col])
                    return (state[0][col] == myPiece) ? 1 : -1;
            }
            // Check diagonal '\'
            if (state[0][0] != ' ' &&
                state[0][0] == state[1][1] && state[1][1] == state[2][2])
                return (state[0][0] == myPiece) ? 1 : -1;
            // Check diagonal '/'
            if (state[0][2] != ' ' &&
                state[0][2] == state[1][1] && state[1][1] == state[2][0])
                return (state[0][2] == myPiece) ? 1 : -1;

            return 0;
        }

        private char[][] deepCopy(char[][] original) {
            char[][] copy = new char[3][3];
            for (int i = 0; i < 3; i++)
                copy[i] = original[i].clone();
            return copy;
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.println("Welcome to Tic Tac Toe Game!");

        TicTacToePlayer ai = new TicTacToePlayer();
        int turn = 0;
        int totalPieces = 0;

        while (ai.gameValue(ai.board) == 0 && totalPieces < 9) {

            // AI's turn
            if (ai.myPiece == TicTacToePlayer.PIECES[turn]) {
                ai.printBoard();
                List<int[]> move = ai.makeMove(ai.board);
                ai.placePiece(move, ai.myPiece);
                System.out.printf("%c moved at (%d, %d)%n",
                                  ai.myPiece, move.get(0)[0], move.get(0)[1]);
                totalPieces++;

            // Human's turn
            } else {
                boolean moveMade = false;
                ai.printBoard();
                System.out.printf("%c's turn%n", ai.opp);

                while (!moveMade) {
                    System.out.print("Move (e.g. row,column): ");
                    String input = scanner.nextLine().trim();
                    String[] parts = input.split(",");

                    if (parts.length != 2) {
                        System.out.println("Invalid format. Use row,column (e.g. 2,3)");
                        continue;
                    }

                    int row, col;
                    try {
                        row = Integer.parseInt(parts[0].trim());
                        col = Integer.parseInt(parts[1].trim());
                    } catch (NumberFormatException e) {
                        System.out.println("Please enter numbers only.");
                        continue;
                    }

                    if (row < 1 || row > 3 || col < 1 || col > 3) {
                        System.out.println("Row and column must be between 1 and 3.");
                        continue;
                    }

                    List<int[]> playerMove = new ArrayList<>();
                    playerMove.add(new int[]{row - 1, col - 1});

                    try {
                        ai.opponentMove(playerMove);
                        moveMade = true;
                    } catch (Exception e) {
                        System.out.println(e.getMessage());
                    }
                }
                totalPieces++;
            }

            turn = (turn + 1) % 2;
        }

        ai.printBoard();
        int result = ai.gameValue(ai.board);
        if (result == 1)       System.out.println("AI wins! Game over.");
        else if (result == -1) System.out.println("You win! Game over.");
        else                   System.out.println("It's a draw! Game over.");

        scanner.close();
    }
}